### PR TITLE
Non-record: AIR-MM v2 Debug Prototype — experimental importance routing

### PR DESCRIPTION
**Pull request by jmccrayiii**
Non-record: AIR-MM v2 Debug Prototype — experimental importance routing

This PR introduces Adaptive Importance Routing Memory Model (AIR-MM) as an experimental non-record submission.

AIR-MM explores selective compute allocation under strict parameter and training-time constraints. Instead of applying uniform updates across all tokens, the model estimates per-token importance and scales residual updates accordingly.

The current prototype includes:
- Learned token-importance gating (v1)
- Positional recency signal
- Lightweight learned combiner (v2)

The baseline architecture is preserved behind a flag to enable clean ablation and comparison.

Local FAST_DEBUG experiments (CPU, short runs) show:
- Stable training behavior
- Monotonic loss decrease
- Small but consistent directional improvements from baseline → v1 → v2

These results suggest the routing mechanism is active and well-behaved, though not yet strongly differentiating token importance at this scale.

This submission is intentionally positioned as an early experimental direction rather than a leaderboard-ready entry. GPU-scale validation and further refinement are pending.

Next steps:
- Run controlled experiments under full 10-minute training constraints
- Increase selectivity and variance of the importance signal
- Evaluate whether selective routing improves parameter efficiency meaningfully

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: AIR-MM v2 Debug Prototype — experimental importance routing

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## Summary PR #1483 ("AIRMM v2 Debug Prototype") introduces an Adaptive Importance Routing mechanism (AIR-MM v2) with a token-level gating module. After a full read of the 1328-line `train_gpt.py`, none of the flagged illegal patterns are present. ## Pattern Checks ### N-gram / hash XOR bug No n-gram hash table, no XOR-based key manipulation. The model is a standard transformer with GQA attention + relu² MLP. CLEAR. ### Pre-Quant TTT (multi-epoch AdamW on val_tokens without score-first) No test-time training of any kind. `val_tokens` is used exclusively inside `eval_val()` (lines 229–289) under `torch.inference_mode()` with `model.eval()`. No gradients, no optimizer steps, no backward pass on val data. CLEAR. ### Score-first-per-chunk TTT (PR #1413 pattern) No TTT chunking, no `is_last_chunk` guard needed — TTT does not exist in this submission. CLEAR. ### HOLD / Scored-region SLOT No scored-region slot exploitation detected. CLEAR. ## What the PR Actually Does **TokenImportanceGate** (lines 632–710): A small learned MLP (dim → hidden_dim → 1) produces a per-token importance scalar in `[min_scale, 1]`. Optionally combined with a positional recency signal (normalized position index). The gate output scales both attention and MLP residual updates inside each transformer block (lines 739–745): ```python imp = self.importance_gate(x).to(dtype=x.dtype) if self.importance_gate is not None else None x = x + (imp * scaled_attn if imp is not None else scaled_attn) x = x + (imp * scaled_mlp if imp is not None else scaled_mlp) ``` This is a straightforward architectural novelty — learned token-level gating on residual contributions during training and inference, with no...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
